# Installing and importing essential packages

In [ ]:
# Don't run this cell if you have already installed
# the packages in requirements.txt
!pip install openpyxl
!pip install pandas
!pip install matplotlib
!pip install numpy

In [1]:
# Lets import the packages 
import os
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

# Setting the filepath and reading from file
Let's use OS module for sorting the filenames so that 
the program works on any machine 

In [16]:
# Getting the current working directory
base_path = os.getcwd()
# Joining our filename with the the current working directory
file_path = os.path.join(base_path, 'app_data.xlsx')

# Setting the name of dataset as the name of file (which means Appendicitis Data)
app_data = pd.read_excel(file_path)

In [17]:
# Let's have a look at the dataset
app_data

Age        BMI     Sex  Height  Weight  Length_of_Stay  \
0    12.680000  16.900000  female   148.0    37.0             3.0   
1    14.100000  31.900000    male   147.0    69.5             2.0   
2    14.140000  23.300000  female   163.0    62.0             4.0   
3    16.370000  20.600000  female   165.0    56.0             3.0   
4    11.080000  16.900000  female   163.0    45.0             3.0   
..         ...        ...     ...     ...     ...             ...   
777  12.413415  25.250476  female   166.5    70.0             4.0   
778  17.092402  20.429418  female   158.0    51.0             6.0   
779  14.992471  19.909972  female   152.0    46.0             4.0   
780   7.195072  14.295549    male   129.3    23.9             5.0   
781  11.509925  18.171441    male   146.5    39.0             4.0   

             Management       Severity Diagnosis_Presumptive        Diagnosis  \
0          conservative  uncomplicated          appendicitis     appendicitis   
1          conservative  uncomplicated          appendicitis  no appendicitis   
2          conservative  uncomplicated          appendicitis  no appendicitis   
3          conservative  uncomplicated          appendicitis  no appendicitis   
4          conservative  uncomplicated          appendicitis     appendicitis   
..                  ...            ...                   ...              ...   
777    primary surgical  uncomplicated          appendicitis     appendicitis   
778  secondary surgical    complicated          appendicitis     appendicitis   
779    primary surgical  uncomplicated          appendicitis     appendicitis   
780    primary surgical  uncomplicated          appendicitis     appendicitis   
781    primary surgical  uncomplicated          appendicitis     appendicitis   

     ...  Abscess_Location  Pathological_Lymph_Nodes Lymph_Nodes_Location  \
0    ...               NaN                       yes                 reUB   
1    ...               NaN                       NaN                  NaN   
2    ...               NaN                       NaN                  NaN   
3    ...               NaN                       yes                 reUB   
4    ...               NaN                       yes                 reUB   
..   ...               ...                       ...                  ...   
777  ...               NaN                       NaN                  NaN   
778  ...           Douglas                       NaN                  NaN   
779  ...               NaN                       NaN                  NaN   
780  ...               NaN                        no                  NaN   
781  ...               NaN                       yes   rechter Unterbauch   

     Bowel_Wall_Thickening Conglomerate_of_Bowel_Loops Ileus Coprostasis  \
0                      NaN                         NaN   NaN         NaN   
1                      NaN                         NaN   NaN         NaN   
2                      NaN                         NaN   NaN         NaN   
3                      NaN                         NaN   NaN         NaN   
4                      NaN                         NaN   NaN         NaN   
..                     ...                         ...   ...         ...   
777                    NaN                         NaN   NaN         NaN   
778                    yes                         yes   NaN         NaN   
779                    NaN                         NaN   NaN         NaN   
780                    yes                         NaN    no         NaN   
781                    NaN                         NaN   yes         NaN   

    Meteorism Enteritis Gynecological_Findings  
0         NaN       NaN                    NaN  
1         yes       NaN                    NaN  
2         yes       yes                    NaN  
3         NaN       yes                    NaN  
4         NaN       yes                    NaN  
..        ...       ...                    ...  
777       NaN       NaN     

# First analysis of the data
Let's see some basic statistics of the dataset

In [18]:
app_data.describe()

Age         BMI      Height      Weight  Length_of_Stay  \
count  781.000000  755.000000  756.000000  779.000000      778.000000   
mean    11.346483   18.906916  148.017460   43.172542        4.284062   
std      3.529979    4.385252   19.732016   17.390984        2.574057   
min      0.000000    7.827983   53.000000    3.960000        1.000000   
25%      9.200000   15.725294  137.000000   29.500000        3.000000   
50%     11.438741   18.062284  149.650000   41.400000        3.000000   
75%     14.099932   21.179011  163.000000   54.000000        5.000000   
max     18.360000   38.156221  192.000000  103.000000       28.000000   

       Alvarado_Score  Paedriatic_Appendicitis_Score  Appendix_Diameter  \
count      730.000000                     730.000000         498.000000   
mean         5.921918                       5.253425           7.762651   
std          2.155972                       1.958456           2.536671   
min          0.000000                       0.000000           2.700000   
25%          4.000000                       4.000000           6.000000   
50%          6.000000                       5.000000           7.500000   
75%          8.000000                       7.000000           9.100000   
max         10.000000                      10.000000          17.000000   

       Body_Temperature   WBC_Count  Neutrophil_Percentage  \
count        775.000000  776.000000             679.000000   
mean          37.404516   12.670683              71.791163   
std            0.903678    5.366525              14.463656   
min           26.900000    2.600000              27.200000   
25%           36.800000    8.200000              61.400000   
50%           37.200000   12.000000              75.500000   
75%           37.900000   16.200000              83.600000   
max           40.200000   37.700000              97.700000   

       Segmented_Neutrophils   RBC_Count  Hemoglobin         RDW  \
count              54.000000  764.000000  764.000000  756.000000   
mean               64.929630    4.799490   13.380497   13.180291   
std                15.085025    0.499012    1.393271    4.538774   
min                32.000000    3.620000    8.200000   11.200000   
25%                54.500000    4.537500   12.600000   12.300000   
50%                64.500000    4.780000   13.300000   12.700000   
75%                77.500000    5.020000   14.000000   13.300000   
max                91.000000   14.000000   36.000000   86.900000   

       Thrombocyte_Count         CRP   US_Number  
count         764.000000  771.000000  760.000000  
mean          285.252618   31.387899  425.515789  
std            72.494373   57.433588  271.585211  
min            91.000000    0.000000    1.000000  
25%           236.000000    1.000000  198.750000  
50%           276.000000    7.000000  398.500000  
75%           330.000000   33.000000  613.250000  
max           708.000000  365.000000  992.000000

In [19]:
# Let's check how many null values are in there in the dataset

# The following command lists the column names in descending order 
# of the data values(i.e. not null values) present in each column
app_data.notnull().sum().sort_values(ascending=False)

Age                                 781
Management                          781
Severity                            781
Sex                                 780
Diagnosis_Presumptive               780
Diagnosis                           780
Weight                              779
US_Performed                        778
Length_of_Stay                      778
Appendix_on_US                      777
WBC_Count                           776
Body_Temperature                    775
Lower_Right_Abd_Pain                774
Nausea                              774
Migratory_Pain                      773
Peritonitis                         773
Loss_of_Appetite                    772
CRP                                 771
Contralateral_Rebound_Tenderness    767
Coughing_Pain                       766
Stool                               765
Hemoglobin                          764
Thrombocyte_Count                   764
RBC_Count                           764
US_Number                           760


## Filtering data
We notice some columns that are sparsely populated.
They might play a crucial role in the decision process (medically speaking).
The [National Library of Medicine](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8426774/#:~:text=In%20a%20literature%2C%20when%20more,provides%20insignificant%20benefit%20(10).) states that more than 10% missing data skews the result

Hence, we will discard any column which has not null values less than 700, which is ~90% of 780 

In [21]:
# Finding the count of null values in the dataset,
# sorting them in descending order and storing the result 
# in a temporary variable
temp = app_data.notnull().sum().sort_values(ascending=False)

# Selecting the column names which have >500 not null values
# (we use lambda function here)
temp = temp.loc[lambda s: s > 700]

# Let's see how many columns we selected
print("Number of columns selected : ",len(temp))
# Also, let's print out which columns we have selected,
# and the count of not null values in each column
print("\nColumns selected and number of values in them:\n",temp)

Number of columns selected :  34

Columns selected and number of values in them:
 Age                                 781
Management                          781
Severity                            781
Sex                                 780
Diagnosis_Presumptive               780
Diagnosis                           780
Weight                              779
US_Performed                        778
Length_of_Stay                      778
Appendix_on_US                      777
WBC_Count                           776
Body_Temperature                    775
Lower_Right_Abd_Pain                774
Nausea                              774
Migratory_Pain                      773
Peritonitis                         773
Loss_of_Appetite                    772
CRP                                 771
Contralateral_Rebound_Tenderness    767
Coughing_Pain                       766
Stool                               765
Hemoglobin                          764
Thrombocyte_Count                   76

In [22]:
# Now slicing the app_data variable and select the columns that are in temp variable  
# We used temp variable since we only needed it to find out which 
# columns to select. After this operation we don't need temp variable
app_data = app_data[list(temp.index)]

# temp.index returns an index object, which is converted to a list using 
# the list() method

# Let's see the resulting dataframe
app_data

Age          Management       Severity     Sex  \
0    12.680000        conservative  uncomplicated  female   
1    14.100000        conservative  uncomplicated    male   
2    14.140000        conservative  uncomplicated  female   
3    16.370000        conservative  uncomplicated  female   
4    11.080000        conservative  uncomplicated  female   
..         ...                 ...            ...     ...   
777  12.413415    primary surgical  uncomplicated  female   
778  17.092402  secondary surgical    complicated  female   
779  14.992471    primary surgical  uncomplicated  female   
780   7.195072    primary surgical  uncomplicated    male   
781  11.509925    primary surgical  uncomplicated    male   

    Diagnosis_Presumptive        Diagnosis  Weight US_Performed  \
0            appendicitis     appendicitis    37.0          yes   
1            appendicitis  no appendicitis    69.5          yes   
2            appendicitis  no appendicitis    62.0          yes   
3            appendicitis  no appendicitis    56.0          yes   
4            appendicitis     appendicitis    45.0          yes   
..                    ...              ...     ...          ...   
777          appendicitis     appendicitis    70.0          yes   
778          appendicitis     appendicitis    51.0          yes   
779          appendicitis     appendicitis    46.0          yes   
780          appendicitis     appendicitis    23.9          yes   
781          appendicitis     appendicitis    39.0          yes   

     Length_of_Stay Appendix_on_US  ...  US_Number   RDW Height        BMI  \
0               3.0            yes  ...      882.0  12.2  148.0  16.900000   
1               2.0             no  ...      883.0  12.7  147.0  31.900000   
2               4.0             no  ...      884.0  12.2  163.0  23.300000   
3               3.0             no  ...      886.0  13.2  165.0  20.600000   
4               3.0            yes  ...      887.0  13.6  163.0  16.900000   
..              ...            ...  ...        ...   ...    ...        ...   
777             4.0            yes  ...      126.0  13.4  166.5  25.250476   
778             6.0             no  ...        NaN  12.8  158.0  20.429418   
779             4.0             no  ...      127.0  12.8  152.0  19.909972   
780             5.0            yes  ...      128.0  12.7  129.3  14.295549   
781             4.0            yes  ...      129.0  12.7  146.5  18.171441   

    Dysuria Psoas_Sign Neutrophilia  Alvarado_Score  \
0        no        yes           no             4.0   
1       yes        yes           no             5.0   
2        no        yes           no             5.0   
3       yes        yes           no             7.0   
4        no        yes           no             5.0   
..      ...        ...          ...             ...   
777      no        yes          yes             8.0   
778      no         no          yes             5.0   
779     yes         no           no             5.0   
780      no         no          yes             9.0   
781      no         no           no             2.0   

    Paedriatic_Appendicitis_Score Free_Fluids  
0                             3.0          no  
1                             4.0          no  
2                             3.0          no  
3                             6.0          no  
4                             6.0          no  
..                            ...         ...  
777                           7.0          no  
778                           3.0          no  
779                           3.0         yes  
780                           8.0         yes  
781                           2.0          no  

[782 rows x 34 columns]

This gives us the columns in which we will impute the values to make the dataset complete

Before moving on, we will check the Age column for any discrepancies,
because we observed (with .describe in the first step) that minimum value in that column is 0.

In [15]:
# Let's inspect the columns with Age < 2
app_data[app_data['Age'] < 2]

Age        BMI   Sex  Height  Weight  Length_of_Stay  \
303  1.727584  17.928215  male    83.5   12.50             3.0   
405  0.850000  18.490000  male   143.0   37.80             3.0   
638  0.533881  20.370610  male   176.0   63.10             3.0   
656  0.000000  17.102915  male   164.0   46.00             6.0   
676  0.035592  14.097544  male    53.0    3.96            12.0   

           Management       Severity Diagnosis_Presumptive        Diagnosis  \
303      conservative  uncomplicated       no appendicitis  no appendicitis   
405      conservative  uncomplicated          appendicitis  no appendicitis   
638  primary surgical  uncomplicated          appendicitis     appendicitis   
656  primary surgical  uncomplicated          appendicitis     appendicitis   
676  primary surgical    complicated          appendicitis     appendicitis   

     ...  Abscess_Location  Pathological_Lymph_Nodes Lymph_Nodes_Location  \
303  ...               NaN                       NaN                  NaN   
405  ...               NaN                       NaN                  NaN   
638  ...               NaN                       NaN                  NaN   
656  ...               NaN                       yes   rechter Unterbauch   
676  ...               NaN                       NaN                  NaN   

     Bowel_Wall_Thickening Conglomerate_of_Bowel_Loops Ileus Coprostasis  \
303                    NaN                         NaN   NaN         NaN   
405                    NaN                         NaN   NaN         NaN   
638                    NaN                         NaN   NaN         NaN   
656                    yes                         NaN   NaN         NaN   
676                    NaN                         NaN   NaN         NaN   

    Meteorism Enteritis Gynecological_Findings  
303       yes       NaN                    NaN  
405       NaN       NaN                    NaN  
638       NaN       NaN                    NaN  
656       NaN       NaN                    NaN  
676       NaN       NaN                    NaN  

[5 rows x 58 columns]

Upon close inspection, we find that rows with index 405 and 638 needs to be fixed since 
these age, weight and height combinations are impossible in real world.
For the sake of simplicity, we assume that the Age column needs fixing